deltacat/compute/compactor/steps/dedupe.py

```diff
@@ -93,25 +90,27 @@ def _union_primary_key_indices(
 
 def _drop_duplicates_by_primary_key_hash(table: pa.Table) -> pa.Table:
     value_to_last_row_idx = {}
-    row_idx = 0
-    pk_op_chunk_iter = zip(
-        sc.pk_hash_column(table).iterchunks(),
-        sc.delta_type_column(table).iterchunks(),
-    )
-    for (pk_chunk, op_chunk) in pk_op_chunk_iter:
-        pk_op_val_iter = zip(
-            pk_chunk.to_numpy(zero_copy_only=False),
-            op_chunk.to_numpy(zero_copy_only=False),
-        )
-        for (pk_val, op_val) in pk_op_val_iter:
-            # operation type is True for `UPSERT` and False for `DELETE`
-            if op_val:
-                # UPSERT this row
-                value_to_last_row_idx[pk_val] = row_idx
-            else:
-                # DELETE this row
-                value_to_last_row_idx.pop(pk_val, None)
-            row_idx += 1
+
+    pk_hash_np = sc.pk_hash_column_np(table)
+    op_type_np = sc.delta_type_column_np(table)
+
+    assert len(pk_hash_np) == len(op_type_np), \
+            "Both pk hash and delta type should have equal length"
+
+    for row_idx in range(len(pk_hash_np)):
```

Comment:
Since `range` will allocate a list of integers equal to the length of pk_hash_np (e.g. maybe millions of integers per invocation?) can we just revert back to the original zipped iterator here instead to conserve some memory and CPU cycles (unless it's actually slower for some reason)?

(We should probably also have a TODO here to delegate this step to C++ eventually - it's just not an observed performance bottleneck today.)

Reply:
Makes sense, I didn't observe any difference in performance. It may have been overshadowed by the dict. Changing to zip to consume less memory.

Reply:
Good point about poor python-native dictionary performance. Maybe an easier short-term TODO to address here would also be to just switch from the current dictionary to NumPy arrays?

Reply:
Makes sense. I have been exploring with numpy `unique` [method](https://numpy.org/doc/stable/reference/generated/numpy.unique.html) with `return_index=True` and mask of `delta_type`.